processor/src/main/java/org/mapstruct/ap/internal/model/source/selector/TypeSelector.java

```diff
@@ -106,24 +111,101 @@ public TypeSelector(TypeFactory typeFactory) {
+     * @param targetType Target type
+     */
     private void addMappingTargetAndTargetTypeBindings(List<ParameterBinding> availableParams, Type targetType) {
-        availableParams.add( ParameterBinding.forMappingTargetBinding( targetType ) );
-        availableParams.add( ParameterBinding.forTargetTypeBinding( typeFactory.classTypeOf( targetType ) ) );
+        boolean mappingTargetAvailable = false;
+        boolean targetTypeAvailable = false;
+
+        // search available parameter bindings if mapping-target and/or target-type is available
+        for ( ParameterBinding pb : availableParams ) {
+            if ( pb.isMappingTarget() ) {
+                mappingTargetAvailable = true;
+            }
+            else if ( pb.isTargetType() ) {
+                targetTypeAvailable = true;
+            }
+        }
+
+        if ( !mappingTargetAvailable ) {
+            availableParams.add( ParameterBinding.forMappingTargetBinding( targetType ) );
+        }
+        if ( !targetTypeAvailable ) {
+            availableParams.add( ParameterBinding.forTargetTypeBinding( typeFactory.classTypeOf( targetType ) ) );
+        }
     }
 
-    private <T extends Method> SelectedMethod<T> getFirstMatchingParameterBinding(Type targetType,
-            SelectedMethod<T> method, List<List<ParameterBinding>> parameterAssignmentVariants) {
+    private <T extends Method> SelectedMethod<T> getMatchingParameterBinding(Type targetType,
+            Method mappingMethod, SelectedMethod<T> method, List<List<ParameterBinding>> parameterAssignmentVariants) {
+
+        List<List<ParameterBinding>> matchingParameterAssignmentVariants = new ArrayList<>(
+            parameterAssignmentVariants
+        );
 
-        for ( List<ParameterBinding> parameterAssignments : parameterAssignmentVariants ) {
-            if ( method.getMethod().matches( extractTypes( parameterAssignments ), targetType ) ) {
-                method.setParameterBindings( parameterAssignments );
-                return method;
+        Method realMethod = method.getMethod();
+
+        // remove all assignment variants that doesn't match the types from the method
+        matchingParameterAssignmentVariants.removeIf(
+            parameterAssignments ->
+                !realMethod.matches( extractTypes( parameterAssignments ), targetType )
+        );
+
+        if ( matchingParameterAssignmentVariants.isEmpty() ) {
+            // no matching variants found
+            return null;
+        }
+        else if ( matchingParameterAssignmentVariants.size() == 1 ) {
+            // we found exactly one set of variants, use this
+            method.setParameterBindings( first( matchingParameterAssignmentVariants ) );
+            return method;
+        }
+
+        // more than one variant matches, try to find one where even the parameter names are matching
+        for ( Iterator<List<ParameterBinding>> it = matchingParameterAssignmentVariants.iterator(); it.hasNext(); ) {
```

Comment:
why use 2 styles of for loops (inner / outer)? Perhaps better to use one. Makes code harder to understand.

Also: perhaps refactor inner loop into a method.

Reply:
I prefer the for-each style (inner loop), but for the outer-loop I had to use the iterator, because entries within the `matchingParameterAssignmentVariants` list should be removed.

But I moved the inner-loop to an own method and switched to `removeIf` like I do it a few lines before.

Fine for you?

Reply:
its fine 👍